### PR TITLE
rgw: delay the RGW process exit until all actice requests have completed

### PR DESCRIPTION
**Pull request by mkogan1**
Configurably (off by default) wait for up to `rgw_exit_timeout_secs` for all outstanding requests
to complete before exiting the RGW process unconditionally.
(new HTTP requests will not be accepted during this time.)

Fixes: https://tracker.ceph.com/issues/66205

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by mkogan1** [comment]
attached below is a screen recording demo of the operation, explanation:

on the bottom pane an `radosgw` is running with `-f` and the new flag `--rgw_graceful_stop=true`
1st in the left pane an 4GB object is PUT with s3cmd, while PUT is proceeding 
next on the top pane radosgw is stopped with SIGTERM
then lastly on the right pane attempt to perform additional PUT is rejected as it should

Peek 2024-05-23 14-56

**Comment by cbodley** [comment]
interesting timing, i'd just proposed https://github.com/ceph/ceph/pull/57640 to change how pause() works for realm reload

**Review by mattbenjamin — APPROVED** [review]
lgtm

**Comment by cbodley** [comment]
> In my QA run, valgrind picked up an issue: `valgrind error: MismatchedFree operator delete[](void*, unsigned long) MonClient::get_monmap_and_config()`. Unfortunately the valgrind yaml file left behind is missing some file names and line numbers. But it does reference `rgw_global_init`, so the error likely originates in RGW.

i saw the same in a wip-cbodley-testing batch. i haven't seen a recent run against main, but i'm guessing it'll show up there too

**Comment by ivancich** [comment]
> > In my QA run, valgrind picked up an issue: `valgrind error: MismatchedFree operator delete[](void*, unsigned long) MonClient::get_monmap_and_config()`. Unfortunately the valgrind yaml file left behind is missing some file names and line numbers. But it does reference `rgw_global_init`, so the error likely originates in RGW.
> 
> i saw the same in a wip-cbodley-testing batch. i haven't seen a recent run against main, but i'm guessing it'll show up there too

Thanks, Casey. I've added a tracker.

**Comment by ivancich** [comment]
@mkogan1 Given @cbodley 's comments, the QA run was successful. Is this ready to merge then? If so, feel free to do so.

**Comment by mkogan1** [comment]
> @mkogan1 Given @cbodley 's comments, the QA run was successful. Is this ready to merge then? If so, feel free to do so.

thank you, apologies, in light of the mentioned above proposed changes in https://github.com/ceph/ceph/pull/57640
refactored to de-couple from the current pause()/unpause() function implementations

**Comment by mkogan1** [comment]
jenkins test make check

File: src/rgw/rgw_asio_frontend.cc
Comment (cbodley) [inline src/rgw/rgw_asio_frontend.cc:1084]: this part is duplicated just below, which shouldn't be necessary. can you just add a single `if (graceful_stop) {` block under that which only calls lock/unlock?

Comment (mkogan1) [reply src/rgw/rgw_asio_frontend.cc]: @cbodley ack, thank you, ultimately the functionality can be distilled to adding a `pause_mutex` lock/unlock at the correct place, please check amend

**Comment by mkogan1** [comment]
@ivancich @cbodley is there a tracker for the valgrind issue? will try to local repro

**Comment by cbodley** [comment]
> @ivancich @cbodley is there a tracker for the valgrind issue? will try to local repro

https://tracker.ceph.com/issues/66336

**Comment by mkogan1** [comment]
jenkins test windows

**Comment by mkogan1** [comment]
jenkins test docs

**Comment by mkogan1** [comment]
jenkins render docs

**Comment by mkogan1** [comment]
for future reference, configuring extra_container_args as below is also required for correct operation in containerized environment:
```
extra_container_args:
  - "--stop-timeout=120"
```
(`120` or the value of `rgw_exit_timeout_secs` config option if it is not default)

**Comment by cbodley** [comment]
@mkogan1 @mattbenjamin last time we discussed this as a group, we'd decided to handle clean shutdown/restart in the layers above rgw. i'm not sure when we decided to change that plan

my position was that rgw's current behavior to close all connections is the correct default. that lets rgw shut down quickly so it can restart and get back to full capacity with respect to `rgw_max_concurrent_requests`

while `rgw_graceful_stop` doesn't change that default, it does add a new constraint that all frontends will have to support. i would _prefer_ that we limit the number of knobs like this in the frontends

that said, this pr has Matt's approval, and i haven't objected to its merge. though if we're promising this feature to users, we'll probably want some form of regression testing to make sure we don't break it

---
separately, there's the issue of realm reloads. i changed this behavior to match the current shutdown behavior in https://github.com/ceph/ceph/pull/57640, but didn't intend for that to merge until we finished discussion here. i think we need to teach realm reload about this `rgw_graceful_stop` flag now too

**Comment by cbodley** [comment]
@SrinivasaBharath can you please link the rgw suite results?

**Comment by cbodley** [comment]
@mkogan1 since this merged prematurely, i opened https://tracker.ceph.com/issues/68352 to track the realm reload issue

**Comment by mkogan1** [comment]
> @mkogan1 since this merged prematurely, i opened https://tracker.ceph.com/issues/68352 to track the realm reload issue

ack, will prioritize work on it asap

**Pull request merged**
